### PR TITLE
testers.testEqualContents: use diffoscope instead of diffing find output

### PR DESCRIPTION
**Pull request by tie**
## Description of changes

Current implementation of `testers.testEqualContents` is borked in some cases (e.g. see 7671505abc42a02bc4821007119a6c29172e838e). This PR changes it to use [diffoscope](https://diffoscope.org) instead of diff‘ing find output to compare file mode 🫠

Tested with `nix build --file . tests.testers.testEqualContents`.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/build-support/testers/test/default.nix [hunk 1]
```diff
@@ -99,117 +99,116 @@ lib.recurseIntoAttrs {
...
+        cat "$log/testBuildFailure.log"
         (
           set -x
-          grep -F -- "Contents must be equal but were not" $log/testBuildFailure.log
-          grep -E -- 'Only in .*-expected: b' $log/testBuildFailure.log
+          grep -F -- "Contents must be equal, but were not!" "$log/testBuildFailure.log" &&
+          grep -E -- '\+\+\+ .*-expected/b/c' "$log/testBuildFailure.log" &&
+          grep -E -- '--- .*-actual/b/c' "$log/testBuildFailure.log" &&
+          grep -F -- "-ACTUAL" "$log/testBuildFailure.log" &&
+          grep -F -- "+EXPECTED" "$log/testBuildFailure.log"
```
Comment (roberth) [inline pkgs/build-support/testers/test/default.nix:205]:
```suggestion
          grep -F -- "Contents must be equal, but were not!" "$log/testBuildFailure.log"
          grep -E -- '\+\+\+ .*-expected/b/c' "$log/testBuildFailure.log"
          grep -E -- '--- .*-actual/b/c' "$log/testBuildFailure.log"
          grep -F -- "-ACTUAL" "$log/testBuildFailure.log"
          grep -F -- "+EXPECTED" "$log/testBuildFailure.log"
```

Aren't we in `set -e` mode, `errexit`?

**Review by roberth — APPROVED** [review]
Nice cleanup! I didn't know it could do that.

File: pkgs/build-support/testers/test/default.nix (hunk 1 above)
Comment (tie) [reply pkgs/build-support/testers/test/default.nix:205]: >Aren't we in `set -e` mode, `errexit`?

And even `inherit_errexit` from stdenv. Bash is being grumpy. This is either a bug or a known quirk, I’m not really sure—`dash` has the same behavior.

- Case 1, works as expected.
  ```bash
  set -e -u
  shopt -s inherit_errexit

  # Case 1.
  (
  	set -x

  	false

  	# OK, this is not reachable.
  	echo unreachable
  )
  ```
  ```
  + false
  ```

- Case 2 (old test). Never trust errexit when pipes or subshells are involved.
  ```bash
  set -e -u
  shopt -s inherit_errexit

  # Case 2.
  (
  	set -x
  	# optionally, explicitly set errexit
  	# (e.g. dash does not support setting inherit_errexit via shopt)
  	set -e

  	false

  	echo FAIL
  ) || {
  	echo oh no
  }
  ```
  ```
  + set -e
  + false
  + echo FAIL
  FAIL
  ```

- Case 3 (updated test).
  ```bash
  set -e -u
  shopt -s inherit_errexit

  # Case 3.
  (
  	set -x
  	false && echo OK
  ) || {
  	echo fails as expected
  }
  ```
  ```
  + false
  fails as expected
  ```

File: pkgs/build-support/testers/test/default.nix [hunk 2]
```diff
@@ -99,117 +99,116 @@ lib.recurseIntoAttrs {
...
-          actual = runCommand "actual" {} ''
-            mkdir -p $out/c
-            echo a >$out/a
-            echo d >$out/c/d
-          '';
-        });
-      } ''
+    unequalExe = testers.testBuildFailure (
+      testers.testEqualContents {
+        assertion = "Different file mod bits are not recognized as equal";
```
Comment (philiptaron) [inline pkgs/build-support/testers/test/default.nix:153]:
```suggestion
        assertion = "Different file mode bits are not recognized as equal";
```

File: pkgs/build-support/testers/test/default.nix [hunk 3]
```diff
@@ -99,117 +99,116 @@ lib.recurseIntoAttrs {
...
+        '';
+        actual = runCommand "actual" { } ''
+          touch -- "$out"
+        '';
+      }
+    );
+
+    unequalExeInDir = testers.testBuildFailure (
+      testers.testEqualContents {
+        assertion = "Different file mod bits are not recognized as equal in directory";
```
Comment (philiptaron) [inline pkgs/build-support/testers/test/default.nix:166]:
```suggestion
        assertion = "Different file mode bits are not recognized as equal in directory";
```

File: pkgs/build-support/testers/test/default.nix [hunk 4]
```diff
@@ -99,117 +99,116 @@ lib.recurseIntoAttrs {
...
+          echo b >"$out/b"
+        '';
+      }
+    );
+
+    fileDiff =
+      let
+        log = testers.testBuildFailure (
+          testers.testEqualContents {
+            assertion = "Different files in are not recognized in equal in directories";
```
Comment (philiptaron) [inline pkgs/build-support/testers/test/default.nix:183]:
```suggestion
            assertion = "Different files are not recognized as equal in subdirectories";
```

**Review by philiptaron — APPROVED** [review]
Nifty.

File: pkgs/build-support/testers/test/default.nix (hunk 1 above)
Comment (tie) [reply pkgs/build-support/testers/test/default.nix]: Added a comment to explain that.

https://github.com/NixOS/nixpkgs/blob/b40d043d5a2a1fb473f1c6c853dac444ceadab62/pkgs/build-support/testers/test/default.nix#L216-L219

**Comment by tie** [comment]
@philiptaron, applied your suggestions, added a few more test cases.

**Review by philiptaron — APPROVED** [review]

**Pull request merged**